File: app/src/main/java/com/voipgrid/vialer/callrecord/CallRecordFragment.java
package com.voipgrid.vialer.callrecord;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.Switch;

<<<<<<< HEAD
import com.voipgrid.vialer.Preferences;
=======
import com.voipgrid.vialer.EmptyView;
>>>>>>> 0be2efa... Refactored all shared preference interactions to be via a User object, rather than having to import ambiguous classes like JsonStorage. This creates a much more fluent, readable api throughout the code.
import com.voipgrid.vialer.R;
import com.voipgrid.vialer.User;
import com.voipgrid.vialer.VialerApplication;
import com.voipgrid.vialer.api.models.CallRecord;
import com.voipgrid.vialer.util.BroadcastReceiverManager;
import com.voipgrid.vialer.util.NetworkUtil;

import java.util.List;

import javax.inject.Inject;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.Unbinder;


public class CallRecordFragment extends Fragment
        implements SwipeRefreshLayout.OnRefreshListener, Observer<PagedList<CallRecord>>, MissedCalls.Callback {

    @Inject CallRecordDataSourceFactory factory;
    @Inject CallRecordAdapter adapter;
    @Inject MissedCalls mMissedCalls;
    @Inject MissedCallsAdapter missedCallsAdapter;
    @Inject NetworkUtil mNetworkUtil;
    @Inject BroadcastReceiverManager mBroadcastReceiverManager;

    @BindView(R.id.swipe_container) SwipeRefreshLayout swipeContainer;
    @BindView(R.id.call_records) RecyclerView mRecyclerView;
    @BindView(R.id.show_missed_calls_only_switch) Switch showMissedCallsOnlySwitch;
    @BindView(R.id.call_records_container) View mCallRecordsContainer;
    @BindView(R.id.call_records_unavailable_view) CallRecordsUnavailableView callRecordsUnavailableView;

    private Unbinder unbinder;
    private NetworkChangeReceiver mNetworkChangeReceiver = new NetworkChangeReceiver();

    private boolean fetchCallsFromEntireAccount = false;

    /**
     * The number of call records to fetch with each HTTP request.
     */
    private static final int CALL_RECORDS_PAGE_SIZE = 50;

    /**
     * Display only the user's calls.
     *
     */
    public static CallRecordFragment mine() {
        return new CallRecordFragment();
    }

    /**
     * Display the user's calls for the entire account.
     *
     */
    public static CallRecordFragment all() {
        CallRecordFragment fragment = new CallRecordFragment();
        fragment.fetchCallsFromEntireAccount = true;
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_call_records, null);
        VialerApplication.get().component().inject(this);
        unbinder = ButterKnife.bind(this, view);

        if (fetchCallsFromEntireAccount) {
            factory.fetchCallsFromEntireAccount();
        }

        return view;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupMissedCallsAdapter();
        setupCallRecordAdapter();
        setupSwipeContainer();
        setupRecyclerView();
        showAppropriateRecords();
        showMissedCallsOnlySwitch.setChecked(User.userPreferences.getDisplayMissedCallsOnly());
        onRefresh();
    }

    private void showAppropriateRecords() {
        mRecyclerView.setAdapter(showMissedCalls() ? missedCallsAdapter : adapter);
    }

    private void setupCallRecordAdapter() {
        PagedList.Config config = new PagedList.Config.Builder().setPageSize(CALL_RECORDS_PAGE_SIZE).build();
        new LivePagedListBuilder(factory, config).build().observe(this, this);
        adapter.setActivity(getActivity());
    }

    private void setupMissedCallsAdapter() {
        missedCallsAdapter.setActivity(getActivity());
    }

    private void setupSwipeContainer() {
        swipeContainer.setColorSchemeColors(getResources().getColor(R.color.color_refresh));
        swipeContainer.setOnRefreshListener(this);
        swipeContainer.post(() -> setRefreshing(true));
    }

    /**
     * Helper method to set refreshing on the swipe container but to avoid NPE if it is null.
     *
     * @param refreshing
     */
    private void setRefreshing(boolean refreshing) {
        if (swipeContainer == null) return;

        swipeContainer.setRefreshing(refreshing);
    }

    private void setupRecyclerView() {
        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(VialerApplication.get());
        mRecyclerView.addItemDecoration(new DividerItemDecoration(getActivity(), LinearLayoutManager.VERTICAL));
        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onRefresh() {
        if (!mNetworkUtil.isOnline()) {
            mCallRecordsContainer.setVisibility(View.GONE);
            callRecordsUnavailableView.noInternetConnection();
            return;
        } else {
            callRecordsUnavailableView.hide();
        }

        mCallRecordsContainer.setVisibility(View.VISIBLE);

        if (showMissedCalls()) {
            mMissedCalls.fetch(this, fetchCallsFromEntireAccount);
            return;
        }

        if (factory == null || factory.getPostLiveData() == null || factory.getPostLiveData().getValue() == null) {
            return;
        }

        factory.getPostLiveData().getValue().invalidate();
    }

    @Override
    public void onChanged(PagedList<CallRecord> callRecords) {
        setRefreshing(false);

        CallRecordDataSource value = factory.getPostLiveData().getValue();

        int code = (value == null ? 500 : value.getLastCode());

        if (!String.valueOf(code).startsWith("2")) {
            handleFailedRequest(code);
            return;
        }

        callRecordsUnavailableView.hide();

        if (callRecords.isEmpty()) {
            callRecordsUnavailableView.unavailable();
            return;
        }

        adapter.submitList(callRecords);
    }

    @Override
    public void setUserVisibleHint(boolean visible) {
        super.setUserVisibleHint(visible);
        if (!visible) return;

        if (showMissedCallsOnlySwitch != null) {
            redrawList();
            showMissedCallsOnlySwitch.setChecked(User.userPreferences.getDisplayMissedCallsOnly());
        }
    }

    /**
     * Force the recycler view to redraw the table, updating the relative timestamps.
     *
     */
    private void redrawList() {
        mRecyclerView.setAdapter(mRecyclerView.getAdapter());
    }

    @Override
    public void missedCallsHaveBeenRetrieved(List<CallRecord> missedCallRecords) {
        if (!isAdded()) return;

        setRefreshing(false);

        if (missedCallRecords.isEmpty()) {
            callRecordsUnavailableView.noMissedCalls();
            return;
        }

        callRecordsUnavailableView.hide();
        missedCallsAdapter.setRecords(missedCallRecords);
    }

    @Override
    public void attemptToRetrieveMissedCallsDidFail(int code) {
        handleFailedRequest(code);
    }

    private void handleFailedRequest(int code) {
        if (!isAdded()) return;

        setRefreshing(false);

        getActivity().runOnUiThread(() -> {
            if (code == 401 || code == 403) {
                callRecordsUnavailableView.permissionsFailed();
            }
            else {
                callRecordsUnavailableView.unavailable();
            }
        });
    }

    private boolean showMissedCalls() {
        return User.userPreferences.getDisplayMissedCallsOnly();
    }

    @OnCheckedChanged(R.id.show_missed_calls_only_switch)
    void missedCallsSwitchWasChanged(CompoundButton missedCallsSwitch, boolean checked) {
        User.userPreferences.setDisplayMissedCallsOnly(checked);
        showAppropriateRecords();
        setRefreshing(true);
        onRefresh();
    }

    @Override
    public void onResume() {
        super.onResume();
        redrawList();
        mBroadcastReceiverManager.registerReceiverViaGlobalBroadcastManager(mNetworkChangeReceiver, "android.net.conn.CONNECTIVITY_CHANGE");
    }

    @Override
    public void onPause() {
        super.onPause();
        mBroadcastReceiverManager.unregisterReceiver(mNetworkChangeReceiver);
    }

    public class NetworkChangeReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            if (mCallRecordsContainer.getVisibility() != View.VISIBLE) {
                onRefresh();
            }
        }
    }
}

File: app/src/main/java/com/voipgrid/vialer/callrecord/CallRecordFragment.kt
package com.voipgrid.vialer.callrecord

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.lifecycle.lifecycleScope
import androidx.paging.PagedList
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.voipgrid.vialer.R
import com.voipgrid.vialer.VialerApplication
import com.voipgrid.vialer.callrecord.database.CallRecordEntity
import com.voipgrid.vialer.callrecord.importing.CallRecordsFetcher
import com.voipgrid.vialer.callrecord.importing.NewCallRecordsImporter
import com.voipgrid.vialer.util.BroadcastReceiverManager
import com.voipgrid.vialer.util.NetworkUtil
import kotlinx.android.synthetic.main.fragment_call_records.*
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject

class CallRecordFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {

    private val callRecordViewModel by lazy {
        activity?.run {
            ViewModelProviders.of(this)[CallRecordViewModel::class.java]
        } ?: throw Exception("No activity found")
    }

    private val adapter = CallRecordAdapter()
    private var type = CallRecordViewModel.Type.ALL_CALLS

    @Inject lateinit var newCallRecordsImporter: NewCallRecordsImporter
    @Inject lateinit var networkUtil: NetworkUtil
    @Inject lateinit var broadcastReceiverManager: BroadcastReceiverManager

    /**
     * Regularly refreshes the data set to keep timestamps relevant.
     *
     */
    private val runnable: Runnable = object : Runnable {
        override fun run() {
            adapter.notifyDataSetChanged()
            Handler().postDelayed(this, REFRESH_TIMEOUT)
        }
    }

    private val networkChangeReceiver = object : BroadcastReceiver() {
        override fun onReceive(p0: Context?, p1: Intent?) {
            onRefresh()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        VialerApplication.get().component().inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? = inflater.inflate(R.layout.fragment_call_records, null)

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        callRecordViewModel.calls.observe(this@CallRecordFragment, Observer<PagedList<CallRecordEntity>> {
            adapter.submitList(it)
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter.activity = activity
        call_records.adapter = adapter
        setupSwipeContainer()
        setupRecyclerView()
        fetchCalls()
        callRecordViewModel.updateDisplayedCallRecords(show_my_calls_only.isChecked, type)
        show_my_calls_only.setOnCheckedChangeListener { _, _ ->
            callRecordViewModel.updateDisplayedCallRecords(show_my_calls_only.isChecked, type)
        }
    }

    override fun onResume() {
        super.onResume()
        broadcastReceiverManager.registerReceiverViaGlobalBroadcastManager(networkChangeReceiver, ConnectivityManager.CONNECTIVITY_ACTION)
        Handler().postDelayed(runnable, REFRESH_TIMEOUT)
        call_records?.scrollToTop()
    }

    override fun onPause() {
        super.onPause()
        broadcastReceiverManager.unregisterReceiver(networkChangeReceiver)
        Handler().removeCallbacks(runnable)
    }

    private fun setupSwipeContainer() {
        swipe_container?.setColorSchemeColors(resources.getColor(R.color.color_refresh))
        swipe_container?.setOnRefreshListener(this)
    }

    private fun setupRecyclerView() {
        val mLayoutManager = LinearLayoutManager(VialerApplication.get())
        activity?.let {
            call_records.apply {
                addItemDecoration(DividerItemDecoration(it, LinearLayoutManager.VERTICAL))
                layoutManager = mLayoutManager
                itemAnimator = DefaultItemAnimator()
            }
        }
    }

    /**
     * Refresh the data being shown by attempting to query the api again,
     * the data displayed will then be automatically based on what is in the database.
     *
     */
    override fun onRefresh() {
        if (!networkUtil.isOnline) {
            swipe_container?.isRefreshing = false
            displayError().noInternetConnection()
            return
        }

        hideError()
        fetchCalls()
    }

    private fun fetchCalls() = lifecycleScope.launch {
        call_records.scrollToTop()
        try {
            newCallRecordsImporter.import()
            hideError()
            call_records.apply {
                scrollToTop()
                delay(200)
                scrollToTop()
            }
        } catch (e: CallRecordsFetcher.PermissionDeniedException) {
            if (e.wasRequestForPersonalCalls) {
                displayError().permissionsFailed()
            }
        } catch (e: Exception) {}
        finally {
            swipe_container?.isRefreshing = false
        }
    }

    /**
     * Display an error, the error to display can be chained on from this method.
     *
     * i.e. displayError().permissionsFailed()
     *
     */
    private fun displayError(): CallRecordsUnavailableView {
        call_records.visibility = View.GONE
        return call_records_unavailable_view
    }

    /**
     * Hide the error, making the call records visible again.
     *
     */
    private fun hideError() {
        call_records.visibility = View.VISIBLE
        call_records_unavailable_view.visibility = View.GONE
    }

    /**
     * Change the type of call records being displayed, this is so an external source can control
     * this fragment.
     *
     */
    fun changeType(type: CallRecordViewModel.Type) {
        this.type = type
        callRecordViewModel.updateDisplayedCallRecords(show_my_calls_only.isChecked, type)
        onRefresh()
    }

    companion object {
        const val REFRESH_TIMEOUT: Long = 5000
    }
}

private fun RecyclerView.scrollToTop() {
    smoothScrollToPosition(0)
}
